panic: runtime error: index out of range

Heyo,

when running bazel run snapshots -- collect -v --bazelrc .bazelrc --out-path "snapshots.json", I get:

```
snapshots: bazel path:       /home/peter/.cache/bazelisk/downloads/bazelbuild/bazel-6.3.2-linux-x86_64/bin/bazel
snapshots: bazelrc path:     /home/peter/nexxiot/nx-arch/codemelt/target_repos/nx0/.bazelrc
snapshots: workspace path:   /home/peter/nexxiot/nx-arch/codemelt/target_repos/nx0
snapshots: query expression: //...
snapshots: out path:         /home/peter/nexxiot/nx-arch/codemelt/target_repos/nx0/snapshots.json
snapshots: collecting digests from //...
snapshots: got 79 build events
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/cognitedata/bazel-snapshots/snapshots/go/pkg/collecter.(*collecter).Collect(0xc0005c7b98?, 0xc0005c7bb8)
	external/com_cognitedata_bazel_snapshots/snapshots/go/pkg/collecter/collecter.go:100 +0xeca
main.(*collectCmd).runCollect(0xc0003996b0, 0x0?, {0xc00046a0e0?, 0x0?, 0x0?})
	external/com_cognitedata_bazel_snapshots/snapshots/go/cmd/snapshots/collect.go:131 +0x385
github.com/spf13/cobra.(*Command).execute(0xc00028cc00, {0xc00046a070, 0x7, 0x7})
	external/com_github_spf13_cobra/command.go:940 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc00028c900)
	external/com_github_spf13_cobra/command.go:1068 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	external/com_github_spf13_cobra/command.go:992
main.Execute({0xc00012c010, 0x8, 0x8})
	external/com_cognitedata_bazel_snapshots/snapshots/go/cmd/snapshots/snapshots.go:12 +0x72
main.main()
	external/com_cognitedata_bazel_snapshots/snapshots/go/cmd/snapshots/main.go:12 +0x79
```

I did put this into `.bazelrc` in addition to my other settings:

`build --legacy_important_outputs`

--- a/.bazelrc
+++ b/.bazelrc
@@ -2,8 +2,4 @@
 build --verbose_failures
 test --test_output=errors
 
-# Docs: Use this to suppress generation of the legacy important_outputs field in the TargetComplete event. important_outputs are required for Bazel to ResultStore/BTX integration.
-# This output is used by bazel-snapshots. This flag will be removed in Bazel 9, and we'll have to update snapshots before then.
-common --legacy_important_outputs
-
 try-import %workspace%/user.bazelrc

--- a/snapshots/go/pkg/bazel/buildevents.go
+++ b/snapshots/go/pkg/bazel/buildevents.go
@@ -4,21 +4,37 @@ package bazel
 
 type BuildEventOutput struct {
 	ID struct {
+		NamedSet struct {
+			ID string `json:"id"`
+		}
 		TargetCompleted struct {
 			Label string `json:"label"`
 		}
 	}
+
+	NamedSetOfFiles NamedSetOfFiles `json:"namedSetOfFiles"`
+
 	Completed struct {
 		Success      bool `json:"success"`
 		OutputGroups []struct {
-			Name string `json:"name"`
+			Name     string `json:"name"`
+			FileSets []struct {
+				ID string `json:"id"`
+			} `json:"fileSets"`
 		} `json:"outputGroup"`
-		ImportantOutput []ImportantOutput `json:"importantOutput"`
 	}
 }
 
-type ImportantOutput struct {
-	Name       string   `json:"name"`
-	URI        string   `json:"uri"`
-	PathPrefix []string `json:"pathPrefix"`
+type NamedSetOfFiles struct {
+	Files    []NamedSetOfFilesFile    `json:"files"`
+	FileSets []NamedSetOfFilesFileSet `json:"fileSets"`
+}
+
+type NamedSetOfFilesFile struct {
+	Name string `json:"name"`
+	URI  string `json:"uri"`
+}
+
+type NamedSetOfFilesFileSet struct {
+	ID string `json:"id"`
 }

--- a/snapshots/go/pkg/collecter/BUILD.bazel
+++ b/snapshots/go/pkg/collecter/BUILD.bazel
@@ -1,4 +1,4 @@
-load("@rules_go//go:def.bzl", "go_library")
+load("@rules_go//go:def.bzl", "go_library", "go_test")
 
 go_library(
     name = "collecter",
@@ -13,5 +13,15 @@ go_library(
         "//snapshots/go/pkg/cache",
         "//snapshots/go/pkg/models",
         "@org_golang_google_grpc//metadata",
+    ],
+)
+
+go_test(
+    name = "collecter_test",
+    srcs = ["collecter_test.go"],
+    embed = [":collecter"],
+    deps = [
+        "//snapshots/go/pkg/bazel",
+        "@com_github_stretchr_testify//assert",
     ],
 )

--- a/snapshots/go/pkg/collecter/collecter.go
+++ b/snapshots/go/pkg/collecter/collecter.go
@@ -80,25 +80,46 @@ func (c *collecter) Collect(args *CollectArgs) (*models.Snapshot, error) {
 		buildEvents = bazelc.BuildEventOutput(ctx, args.BazelRcPath, bazelArgs...)
 	}
 
-	// create a map from label to file
-	labelFiles := map[string]string{}
+	bazelFiles := make(namedSetsOfFiles)
+	labelFiles := make(map[string]string) // label -> uri
 	for event, err := range buildEvents {
 		if err != nil {
 			return nil, fmt.Errorf("error reading build event: %w", err)
 		}
+		switch {
+		case event.ID.NamedSet.ID != "":
+			bazelFiles.Put(event.ID.NamedSet.ID, event.NamedSetOfFiles)
 
-		label := event.ID.TargetCompleted.Label
-		var uri string
+		case event.ID.TargetCompleted.Label != "":
+			label := event.ID.TargetCompleted.Label
+			var labelURI string
+		uriSearch:
+			for _, group := range event.Completed.OutputGroups {
+				if group.Name != "change_track_files" {
+					continue
+				}
 
-		for idx, g := range event.Completed.OutputGroups {
-			if g.Name == "change_track_files" {
-				uri = event.Completed.ImportantOutput[idx].URI
-				break
+				// Found a change_track_files output group.
+				// Find the file set for this label.
+				// File sets can point to other file sets,
+				// so we need to traverse them
+				// until we find the actual change tracker files.
+				//
+				// Bazel guarantees that a file set will be
+				// provided before it's referenced.
+				for _, fileSet := range group.FileSets {
+					for uri := range bazelFiles.ByID(fileSet.ID) {
+						// change_track_files will
+						// contain only one file per label,
+						// so we can stop at the first one.
+						labelURI = uri
+						break uriSearch
+					}
+				}
 			}
-		}
-
-		if label != "" && uri != "" {
-			labelFiles[label] = uri
+			if label != "" && labelURI != "" {
+				labelFiles[label] = labelURI
+			}
 		}
 	}
 	log.Printf("got %d change trackers", len(labelFiles))
@@ -153,6 +174,51 @@ func (c *collecter) Collect(args *CollectArgs) (*models.Snapshot, error) {
 	}
 
 	return manifest, nil
+}
+
+// namedSetsOfFiles is an in-memory buffer for NamedSetOfFiles.
+//
+// Bazel produces NamedSetOfFiles events in the build event stream,
+// alongside TargetCompleted events.
+// Both events can reference other NamedSetOfFiles by ID.
+// Bazel guarantees that a NamedSetOfFiles event will be
+// provided before it's referenced.
+//
+// To use namedSetsOfFiles, populate it with NamedSetOfFiles events
+// as they are produced by Bazel using Put().
+//
+// To retrieve all files in a NamedSetOfFiles by ID,
+// use ByID(), to get an iterator over the file URIs.
+type namedSetsOfFiles map[string]bazel.NamedSetOfFiles
+
+func (fs namedSetsOfFiles) Put(id string, ns bazel.NamedSetOfFiles) {
+	fs[id] = ns
+}
+
+func (fs namedSetsOfFiles) ByID(fileSetID string) iter.Seq[string] {
+	return func(yield func(string) bool) {
+		pending := []string{fileSetID}
+		seen := make(map[string]struct{})
+		for len(pending) > 0 {
+			currentID := pending[len(pending)-1]
+			pending = pending[:len(pending)-1]
+
+			if _, ok := seen[currentID]; ok {
+				continue // already seen this file set
+			}
+			seen[currentID] = struct{}{}
+
+			for _, file := range fs[currentID].Files {
+				if !yield(file.URI) {
+					return
+				}
+			}
+
+			for _, fileSet := range fs[currentID].FileSets {
+				pending = append(pending, fileSet.ID)
+			}
+		}
+	}
 }
 
 func createMetadata(input []string) metadata.MD {

--- a/snapshots/go/pkg/collecter/collecter_test.go
+++ b/snapshots/go/pkg/collecter/collecter_test.go
@@ -1,0 +1,110 @@
+package collecter
+
+import (
+	"slices"
+	"testing"
+
+	"github.com/cognitedata/bazel-snapshots/snapshots/go/pkg/bazel"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNamedSetOfFiles(t *testing.T) {
+	type fileSetEvent struct {
+		ID    string
+		Files bazel.NamedSetOfFiles
+	}
+
+	tests := []struct {
+		name string
+		give []fileSetEvent
+		want map[string][]string // id -> []files
+	}{
+		{
+			name: "UnknownFileSet",
+			want: map[string][]string{
+				"0": {},
+			},
+		},
+		{
+			name: "FileSetWithFiles",
+			give: []fileSetEvent{
+				{
+					ID: "0",
+					Files: bazel.NamedSetOfFiles{
+						Files: []bazel.NamedSetOfFilesFile{
+							{Name: "a", URI: "file:///a"},
+							{Name: "b", URI: "file:///b"},
+						},
+					},
+				},
+			},
+			want: map[string][]string{
+				"0": {"file:///a", "file:///b"},
+			},
+		},
+		{
+			name: "FileSetWithNestedFileSets",
+			give: []fileSetEvent{
+				{
+					ID: "0",
+					Files: bazel.NamedSetOfFiles{
+						Files: []bazel.NamedSetOfFilesFile{
+							{Name: "a", URI: "file:///a"},
+							{Name: "b", URI: "file:///b"},
+						},
+					},
+				},
+				{
+					ID: "1",
+					Files: bazel.NamedSetOfFiles{
+						Files: []bazel.NamedSetOfFilesFile{
+							{Name: "c", URI: "file:///c"},
+						},
+						FileSets: []bazel.NamedSetOfFilesFileSet{
+							{ID: "0"},
+						},
+					},
+				},
+				{
+					ID: "2",
+					Files: bazel.NamedSetOfFiles{
+						Files: []bazel.NamedSetOfFilesFile{
+							{Name: "d", URI: "file:///d"},
+						},
+					},
+				},
+				{
+					ID: "3",
+					Files: bazel.NamedSetOfFiles{
+						FileSets: []bazel.NamedSetOfFilesFileSet{
+							{ID: "1"},
+							{ID: "2"},
+						},
+					},
+				},
+			},
+			want: map[string][]string{
+				"0": {"file:///a", "file:///b"},
+				"1": {"file:///c", "file:///a", "file:///b"},
+				"2": {"file:///d"},
+				"3": {"file:///c", "file:///a", "file:///b", "file:///d"},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			files := make(namedSetsOfFiles)
+			for _, ev := range tt.give {
+				files.Put(ev.ID, ev.Files)
+			}
+
+			for id, wantFiles := range tt.want {
+				got := slices.Collect(files.ByID(id))
+
+				// It's a set so file order is not guaranteed.
+				assert.ElementsMatch(t, wantFiles, got, "files for id %q", id)
+			}
+		})
+	}
+}

